### PR TITLE
Various quad-buffered ("HDMI 3D") stereo fixes

### PR DESCRIPTION
**Pull request by stenzek**
Fixes:
- Fixes the same image was rendered to both eyes in quad-buffered modes in D3D/Vulkan modes.
- Removes the redundant geometry shader in OpenGL mode, as the eyes are rendered separately.

New:
- Renders the UI (ImGui) to both eyes. Previously, this was only rendered to the left eye, which I imagine would look incorrect if using a HMD.

I've only tested D3D and Vulkan by checking both layers in RenderDoc, as I don't have a compatible display to check against. GL just gives a `GL_INVALID_OPERATION` result when selecting the eyes via `glDrawBuffers`, which is expected without a display.

### CODE REVIEW

**Comment by iwubcode** [comment]
Why do we only want ImGui to be rendered to both eyes if using a quad buffered stereo?  Wouldn't this also cause problems in other stereo solutions?

**Comment by mbc07** [comment]
Is this PR stalled due the lack of testing? I still own a 2012 Sony BRAVIA 3D TV I can probably use to test this PR, just let me know...

**Comment by jordan-woyak** [comment]
@mbc07 Testing that both eyes work in HDMI 3D would be great.

**Comment by mbc07** [comment]
So, I tested this a bit and unfortunately couldn't draw any conclusion. The only PC I could hook to the TV is old and thus it also limited the video backends I could test:

**CPU:** Pentium G3258 @ 4.2 GHz (overclocked)
**RAM:** 16 GB (4x4 GB, DDR3-1600)
**iGPU:** Intel HD Graphics (Haswell GT1), driver version: 15.40.42.5058
**dGPU**: NVIDIA GeForce GT 440 (1 GB DDR3), driver version: 391.35
**OS:** Windows 10 v1903 (build 18362.418)

- **3D TV model:** Sony BRAVIA KDL-40HX755 
- **Dolphin versions tested:** 5.0-11077 and latest build of this PR. 
- **Games tested:** Luigi's Mansion (Intro Scene) and Zelda Twilight Princess (Start Screen Scene)

All these builds were tested in portable mode and all settings were the default ones, except for the video backend (could test only DX11 and OpenGL) and the stereoscopy mode (changed to HDMI 3D). Also enabled "Show FPS" to test ImGui rendering and "Use fullscreen" to make sure Dolphin were running in exclusive fullscreen mode.

**Test 1:** NVIDIA GPU, Dolphin 5.0-11077:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, only thing displayed was a solid color that seemed to change depending of what the game was trying to render. FPS and OSD messages were displayed correctly. Screenshots captured with Dolphin showed the game output just fine, despite it rendering only a solid color on the screen.
* **Direct3D 11:** at boot, no errors or warnings. TV stayed in 2D mode, game displayed normally on screen (although in 2D only), FPS and OSD messages were displayed correctly. Upon stopping emulation, Dolphin spit some errors (in this order):
dx11-master-error1
dx11-master-error2
dx11-master-error3
dx11-master-error4

**Test 2:** Intel GPU, Dolphin 5.0-11077:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, only thing displayed was a solid color that seemed to change depending of what the game was trying to render. FPS and OSD messages were displayed correctly. Screenshots captured with Dolphin showed the game output just fine, despite it rendering only a solid color on the screen.
* **Direct3D 11:** no errors or warnings, and this time the TV switched to 3D mode ("Detected a 3D Signal" shown on-screen), However, the game was still at 2D, the right eye image looked identical to the left eye image and randomly got swapped with a identical frame but slighlty offset upwards. FPS and OSD messages appeared only on the left eye and looking at them for prolonged time caused some eye strain.

**Test 3:** NVIDIA GPU, PR #8378:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui also rendered normally.
* **Direct3D 11:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui also rendered normally.

**Test 4:** Intel GPU, PR #8378:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui also rendered normally.
* **Direct3D 11:** at boot, Dolphin spit the error below. Upon clicking on OK, the TV switched to 3D mode ("Detected a 3D Signal" shown on-screen) but the game rendered in 2D and looked darker overall. The left eye contained the game screen and ImGui output, while the right eye was just a black frame with the ImGui output (confirmed this by blocking my left eye with my hand -- that probably explains the darkened look of the image when looking with both eyes).
dx11-pr8352-error

**Test 5 (bonus):** NVIDIA GPU, PR #8378, NVIDIA 3D Vision mode:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui also rendered normally.
* **Direct3D 11:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui also rendered normally.

Unfortunately I only noticed the "Passive" 3D mode (which from the description seems aimed specifically at TVs) after I finished moving the PC back to my room, so I didn't test it in any of the builds. Anyway, just let me know if you want this tested too, I can do that a later time...

**TL;DR** quad-buffered stereo mode didn't work in any of the tested builds, although it seemed less buggy after this PR, especially with OpenGL. Using Side-by-side or Top-and-bottom modes worked perfectly in all tested builds, with both backends and both GPUs.

Also, ImGui being rendered to both eyes is a very welcome improvement and it would be great if this change could be "ported" to all other stereoscopy modes...

**Comment by jordan-woyak** [comment]
@mbc07 Thanks for testing.

Just an FYI, this reportedly fixes stereo output in my OpenXR PR #8380.

**Comment by stenzek** [comment]
@mbc07 thanks for the testing. It's possible something else has to be done apart from just placing the swap chain in stereo mode, unfortunately I'm not the original author of the implementation so it's difficult to say if that's the issue or if something else caused it to regress at some point.

The right eye being black is a little concerning though - that should've been fixed by this PR.

**Comment by mbc07** [comment]
> The right eye being black is a little concerning though - that should've been fixed by this PR.

Perhaps it was black because of the DX11 error that popped at boot? (Create integer render target view for framebuffer failed)

**Comment by mbc07** [comment]
@stenzek I have reasons to think my previous test results are invalid: just noticed that sometimes a NVIDIA overlay saying that "no 3D Vision Emitter was detected" briefly appears on 3D applications running on that PC even when not connected to a 3D screen and also noticed that trying to play 3D Blu-rays with PowerDVD 18 (which I'm 100% sure it works as I watched lots of 3D movies before with that setup) is also failing and displaying only in 2D now.

So, I'm gonna do a clean install of the GPU drivers and test this again, as something definitely is wrong. Will also throw 5.0-4657 (the first Dolphin version with quad-buffered stereo) and 5.0-5380 (which include some important fixes) on the test too...

**Comment by stenzek** [comment]
@mbc07 ah, interesting. It's also strange that it only switches to 3D mode on the Intel GPU, the rendering being broken without this PR does make sense, but the fact that only the UI is being displayed on the right eye suggests something might be wrong on our end. I'll double-check the presentation path when I get a chance to make sure it is actually drawing both layers..

**Comment by mbc07** [comment]
@stenzek I retested today and indeed the results were very different. This time, there were no errors at all on any of the tested builds and, while the NVIDIA GPU still didn't switch the TV to 3D mode, seeing the screen refresh rate now changing at boot is a good indicator that maybe we're just missing a small detail somewhere or that's just an issue with the legacy drivers that might not occur with the current NVIDIA drivers.

So, I used DDU in safe mode to completely wipe both Intel and NVIDIA drivers from the PC, then proceeded installing driver 391.35 for NVIDIA (last version for Fermi GPUs before EOL) and driver 15.40.42.5063 for Intel (last version for Haswell iGPUs before EOL).

This time the tested builds were **5.0-5380**, **5.0-11077** (for comparison with the previous tests) and this PR. I ended skipping **5.0-4657** because the only differences regarding quad-buffered stereo between it and **5.0-5380** is that the latter switches to quad-buffered mode only when "HDMI 3D" is selected while the former always switches to this mode if the connected display supported it, even if stereoscopy is currently disabled in Dolphin. So, the new results:

**Test 1:** NVIDIA GPU, Dolphin 5.0-5380:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. FPS and OSD Messages also rendered normally.
* **Direct3D 11:** no errors, warnings, nothing. Screen refresh rate switched to 30 Hz at boot but the TV stayed in 2D mode regardless. Game rendered in 2D, normally, FPS and OSD messages also rendered normally. Upon closing Dolphin, screen refresh rate switched back to 60 Hz.

**Test 2:** Intel GPU, Dolphin 5.0-5380:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. FPS and OSD Messages also rendered normally.
* **Direct3D 11:** no errors, warnings, nothing. TV switched to 3D mode ("Detected a 3D Signal" shown on-screen) and game rendered in 3D, normally (see note about possible driver bug below). FPS and OSD messages rendered only on the left eye.

**Test 3:** NVIDIA GPU, Dolphin 5.0-11077:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, only thing displayed was a solid color that seemed to change depending of what the game was trying to render. ImGui output rendered correctly. Screenshots captured with Dolphin showed the game output just fine, despite it rendering only a solid color on the screen.
* **Direct3D 11:** no errors, warnings, nothing. Screen refresh rate switched to 30 Hz at boot but the TV stayed in 2D mode regardless. Game rendered in 2D, normally, and ImGui output also rendered normally. Upon closing Dolphin, screen refresh rate switched back to 60 Hz.

**Test 4:** Intel GPU, Dolphin 5.0-11077:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, only thing displayed was a solid color that seemed to change depending of what the game was trying to render. ImGui output rendered correctly. Screenshots captured with Dolphin showed the game output just fine, despite it rendering only a solid color on the screen.
* **Direct3D 11:** no errors, warnings, nothing. TV switched to 3D mode ("Detected a 3D Signal" shown on-screen) but the game still rendered in 2D as exactly the same frame was displayed on both eyes (see note about possible driver bug below). ImGui output rendered only on the left eye.

**Test 5:** NVIDIA GPU, this PR build:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui output rendered normally.
* **Direct3D 11:** no errors, warnings, nothing. Screen refresh rate switched to 30 Hz at boot but the TV stayed in 2D mode regardless. Game rendered in 2D, normally, and ImGui output also rendered normally. Upon closing Dolphin, screen refresh rate switched back to 60 Hz.

**Test 6:** Intel GPU, this PR build:
* **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui output rendered normally.
* **Direct3D 11:** no errors, warnings, nothing. TV switched to 3D mode ("Detected a 3D Signal" shown on-screen) and game rendered in 3D, normally (see note about possible driver bug below). ImGui output  rendered to both eyes.

Like before, Side-by-side and Top-and-bottom stereoscopy modes worked correctly in all backends on all tested builds. I already said that before but I'll reiterate that ImGui rendering on both eyes (added on this PR but only to quad-buffered mode) is a very welcome improvement and it would be nice if it could be ported to the other stereoscopy modes too.

**Possible Intel driver bug:** as evident on the new tests, regardless of the build tested, using DX11 with the Intel GPU was the only way for the TV to "notice" it was receiving a quad-buffered stereoscopy signal and make it switch to 3D mode. However, I consistently hit a bug which is very unlikely to be caused by Dolphin as once triggered, it persisted until rebooting the PC. It's also very unlikely to be fixed as Haswell iGPUs hit EOL a long time ago. 

Anyway, once triggered, the image of the right eye will keep constantly flickering between the correct frame and a identical but slightly offset upwards frame, [like on this video](https://imgur.com/HaeEoeC). It happens around 90% of time if you set Dolphin to boot the games in exclusive full screen and way less (I would say 20% of time) if you set Dolphin to boot the games in windowed mode or in borderless full screen. Once triggered, the bug keeps happening even after closing Dolphin and will also affect anything being displayed on screen, as the TV itself is kept in 3D mode by the driver. Only rebooting the PC will fix that and also "release" the TV to go back to 2D mode.

**TL;DR** quad-buffered stereoscopy works fine on the Intel GPU with DX11 backend. Dolphin 5.0-5380 renders correctly, current development builds are broken but this PR fixes that as the games renders correctly in 3D once again on it, with the benefit of now having ImGui rendering on both eyes...

**Comment by stenzek** [comment]
@mbc07 thank you for the detailed testing results! I think this PR is good to go then. Adding the UI rendering to both eyes should be a simple change.

It's a shame it doesn't work on NVIDIA for whatever reason.

**Comment by ThreeDeeJay** [comment]
> **Test 5 (bonus):** NVIDIA GPU, PR #8378, NVIDIA 3D Vision mode:
> 
> - **OpenGL:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui also rendered normally.
> 
> - **Direct3D 11:** no errors, warnings, nothing. TV stayed in 2D mode, game rendered in 2D, normally. ImGui also rendered normally.

> It's a shame it doesn't work on NVIDIA for whatever reason.


@stenzek @mbc07 Kinda (quite, actually) late to the party, but I can confirm HDMI 3D works in 3D Vision (even in a window) even in the latest Dolphin release, though I think it requires being on the latest official 3DV driver <=425.31 (so only <=2080 Ti) which I'm currently on by no coincidence. Direct3D 11, 12 and OpenGL all worked, though the latter didn't have great performance.
